contract/models/account_analytic_account.py

```diff
@@ -259,38 +266,52 @@ def _create_invoice(self):
         invoice.compute_taxes()
         return invoice
 
+    @contextmanager
+    @api.multi
+    def logged_exception_savepoint(self, *args):
+        try:
+            with self.env.cr.savepoint() as ctx:
+                yield ctx
+        except:
+            _logger.exception(traceback.format_exc())
+            if args:
+                _logger.exception(*args)
+
     @api.multi
     def recurring_create_invoice(self):
         """Create invoices from contracts
 
         :return: invoices created
         """
         invoices = self.env['account.invoice']
+
+        error_msg_tmpl = ('Invoice creation failed for contract id %d'
+                          ' (traceback above).'
+                          ' Going on with other contracts, if any.')
         for contract in self:
-            ref_date = contract.recurring_next_date or fields.Date.today()
-            if (contract.date_start > ref_date or
-                    contract.date_end and contract.date_end < ref_date):
-                if self.env.context.get('cron'):
-                    continue  # Don't fail on cron jobs
-                raise ValidationError(
-                    _("You must review start and end dates!\n%s") %
-                    contract.name
-                )
-            old_date = fields.Date.from_string(ref_date)
-            new_date = old_date + self.get_relative_delta(
-                contract.recurring_rule_type, contract.recurring_interval)
-            ctx = self.env.context.copy()
-            ctx.update({
-                'old_date': old_date,
-                'next_date': new_date,
-                # Force company for correct evaluation of domain access rules
-                'force_company': contract.company_id.id,
-            })
-            # Re-read contract with correct company
-            invoices |= contract.with_context(ctx)._create_invoice()
-            contract.write({
-                'recurring_next_date': fields.Date.to_string(new_date)
-            })
+            with self.logged_exception_savepoint(error_msg_tmpl, contract.id):
```

Comment:
Is it needed to enclose this in a new method? When any errors happen, you will already have logging entries, so with this, you will duplicate the logging activity.

Reply:
After digging a bit more into the logging issue @pedrobaeza , I realized that logging the traceback is required here, otherwise the exception will never get logged as it is blocked in the logged_exception_savepoint and will not make it to the catch-all except clause of the kernel.

This is illustrated by the contract module tests that use the new assertExceptionLogged test method: there is no raised exception anymore, so the previous assertRaises call has to be replaced by an error logging check.

Reply:
OK, but please enclose on that only the strict called to `create_invoice` and let the rest out of it, so the diff is lower and the rest of possible issues are not masked.

Reply:
Hello again @pedrobaeza and thanks for reviewing this again and again.

I first intended to do just that (even before you suggested), but changed my mind because:
* if it were only a matter of diff viewing, it would be a bad thing; git diff has a "-w" option that serves exactly this purpose, and it has a counterpart in the github UI ("Diff settings > Hide whitespace changes"); this really makes the change easy to read
* the code is clearer and of better quality as is IMHO, because the whole idea is to isolate contract handling: a crash wrt. a given contract should have no influence on the others; the present implementation does exactly that, with the context manager enclosing the whole loop content; if we do not do that, not matter the error before the context manager, it would rollback the whole transaction and destroy the desired effect.

Reply:
It's not a question of viewing, but commit diff size. Any way, if you want to do that, then extract code to an isolated method, so that we can better read it, hook into it, etc